### ISSUE TITLE
Geometry and activation state of levers

### ISSUE THREAD

**Issue by Svarr** [RavinMaddHatter/Structura#153]
Currently, the ghost-block levers' models don't represent their activation state (`open_bit`). I'm currently working on changing the geometry definition in `block_shapes.json`, so that the stick-part of the lever is not perpendicular to the base-part anymore. The only way I found, was to add an optional json property to represent rotation angles for each part of the shape and apply these in the `armorstandgeo.make_block` function.

This, however, leads to issues with the block rotations specified in `block_rotation.json`, as they all are euler rotations (since rotations of 90° are very common in block rotations, there is an especially high risk of gimbal lock). I couldn't find any information about how and when the rotations defined in `block_rotation.json` are actually applied to the shapes.

The commit https://github.com/Svarr/Structura/commit/4f0bb08ce9c1d97fbdb3550b7bea4698cbf6b1de shows the current state of my progress on this.

There are a few issues specific to this implementation:
- the lever stick is rotated around the wrong origin (its center, should be the base) -> proper fix would be to allow for providing pivot points per shape and not per block, hacky fix would be to fiddle with the offset values until it looks decent enough
- some of the rotations seem to be wrong (most notably, the stick is upside-down for 2 of the `lever_direction` values, which I assume to be caused by gimbal lock)

**Comment by Svarr** [comment]
> The refactor is in the layer above this. So I haven't changed anything in the actual block code. If there is a recommendation on how to change the block code I can look at that.
>
> Euler angles are a place I am admittedly weak. Do you have any suggestions on how I would change this in general, I want to avoid changes that are block specific if possible. But just how you would describe a block genetically would help.

_Originally posted by @RavinMaddHatter in https://github.com/RavinMaddHatter/Structura/issues/152#issuecomment-1407498414_

The best solution I could come up with, you can find in my fork (see link in my post above). I was aiming for an implementation that is as universally usable (i.e. by other blocks) as possible, while not requiring to touch the already existing data at all. I already explained the details above and in PR #152.

Any block that features 'diagonal' geometry (apart from the lever I can't think of any from the top of my head, though) can add the `rotation` property to it's `block_shapes`-definition as shown [here (L298 of my fork's `block_shapes.json`)](https://github.com/Svarr/Structura/blob/4f0bb08ce9c1d97fbdb3550b7bea4698cbf6b1de/lookups/block_shapes.json#L298). These values are then applied by the code if present (see [armor_stand_geo_class_2.py#L141](https://github.com/Svarr/Structura/blob/4f0bb08ce9c1d97fbdb3550b7bea4698cbf6b1de/armor_stand_geo_class_2.py#L141)).

Regarding the gimbal lock issues, it depends on how the geometry data needs to be fed to Minecraft. If you **have** to apply multiple rotations to the same object, quaternions would be your friend. But this wouldn't be necessary in this case: the rotation defined in `block_rotation.json` should be applied to the whole block, while the rotations that can be defined in `block_shapes.json` should only be applied to that part of the geometry it was defined for (addressed by index, just like with "size" and "offsets"). So the hierarchical approach, which solves the issue regarding gimbal lock, would be the way to go. I checked the documentation for `.geo.json`-files and found that you can indeed define a rotation for the whole bone. I'll change the code to use that and see if that works.

**Comment by Svarr** [comment]
I changed the code to apply the block rotation to the whole bone and only the rotations defined in `block_shapes.json` to its corresponding children and it did the trick. The branch [fix/lever-geo](https://github.com/Svarr/Structura/tree/fix/lever-geo) now fixes the lever geometry, rotation and activated state, as per commit https://github.com/Svarr/Structura/commit/9a177365209925223c32a5b12c877203121474bd.

The only issue I encountered is that the block requirements list is not created (even if the checkbox is ticked). But I haven't checked yet, whether that's been caused by my changes. I'll create a PR once I did.

**Comment by RavinMaddHatter** [comment]
I have a had a crazy few weeks. sorry about not getting back to you. I will work on getting this merged in soon

**Comment by Svarr** [comment]
No worries, I've got things going on myself...

I just checked why the block list isn't created and the cause appears to be `runFromGui()` only calling the `generate_nametag_file()` function if the checkbox for the variable `export_list` is ticked. The functions `make_big_blocklist()` and `make_nametag_block_lists()` don't seem to be called at all. This may be resolved in your refactoring, though.

Anyway, I have created the PR #156. I'll see if I can merge your recent changes into it.
